cpp/oneapi/dal/backend/primitives/objective_function/logloss_functors_dpc.cpp

```diff
@@ -0,0 +1,532 @@
+          n_(data.get_row_count()),
+          p_(data.get_column_count()),
+          bsz_(bsz == -1 ? get_block_size(n_, p_) : bsz) {
+    raw_hessian_ = ndarray<Float, 1>::empty(q_, { n_ }, sycl::usm::alloc::device);
+    buffer_ = ndarray<Float, 1>::empty(q_, { n_ }, sycl::usm::alloc::device);
+    tmp_gpu_ = ndarray<Float, 1>::empty(q_, { p_ + 1 }, sycl::usm::alloc::device);
+    if (data_.get_kind() == dal::csr_table::kind()) {
+        sp_handle_.reset(new sparse_matrix_handle(q));
+        set_csr_data(q_, *sp_handle_, static_cast<const csr_table&>(data_));
+    }
+}
+
+template <typename Float>
+logloss_hessian_product<Float>::logloss_hessian_product(sycl::queue& q,
+                                                        comm_t comm,
+                                                        const table& data,
+                                                        Float L2,
+                                                        bool fit_intercept,
+                                                        std::int64_t bsz)
+        : q_(q),
+          comm_(comm),
+          data_(data),
+          L2_(L2),
+          fit_intercept_(fit_intercept),
+          n_(data.get_row_count()),
+          p_(data.get_column_count()),
+          bsz_(bsz == -1 ? get_block_size(n_, p_) : bsz) {
+    raw_hessian_ = ndarray<Float, 1>::empty(q_, { n_ }, sycl::usm::alloc::device);
+    buffer_ = ndarray<Float, 1>::empty(q_, { n_ }, sycl::usm::alloc::device);
+    tmp_gpu_ = ndarray<Float, 1>::empty(q_, { p_ + 1 }, sycl::usm::alloc::device);
+    if (data_.get_kind() == dal::csr_table::kind()) {
+        sp_handle_.reset(new sparse_matrix_handle(q));
+        set_csr_data(q_, *sp_handle_, static_cast<const csr_table&>(data_));
+    }
+}
+
+template <typename Float>
+ndview<Float, 1>& logloss_hessian_product<Float>::get_raw_hessian() {
+    return raw_hessian_;
+}
+
+template <typename Float>
+sycl::event logloss_hessian_product<Float>::compute_with_fit_intercept(const ndview<Float, 1>& vec,
+                                                                       ndview<Float, 1>& out,
+                                                                       const event_vector& deps) {
+    ONEDAL_PROFILER_TASK(compute_hessp_with_fit_intercept, q_);
+    auto* const tmp_ptr = tmp_gpu_.get_mutable_data();
+    ONEDAL_ASSERT(vec.get_dimension(0) == p_ + 1);
+    ONEDAL_ASSERT(out.get_dimension(0) == p_ + 1);
+    auto fill_buffer_event = fill<Float>(q_, buffer_, Float(1), deps);
+    auto out_suf = out.get_slice(1, p_ + 1);
+    auto tmp_suf = tmp_gpu_.slice(1, p_);
+    auto out_bias = out.get_slice(0, 1);
+    auto vec_suf = vec.get_slice(1, p_ + 1);
+    ndview<Float, 1> tmp_ndview = tmp_gpu_;
+
+    sycl::event fill_out_event = fill<Float>(q_, out, Float(0), deps);
+
+    Float v0 = vec.at_device(q_, 0, deps);
+    event_vector last_iter_deps = { fill_buffer_event, fill_out_event };
+
+    if (data_.get_kind() == dal::csr_table::kind()) {
```

Comment:
Please consider moving the body of these two branches into a separate functions/